### DISCUSSION TITLE
gcc

### DISCUSSION THREAD

**Discussion by Selmacas** [SpenceKonde/megaTinyCore#667] [Q&A]
Hello, i have some attiny3226 and i want to develop for them. I've found that avr-gcc 5.4.0 from system repositories doesn't support them and xc8 compiller is not possible to install in raspberry pi. The only way to compile for them is to use avr-gcc installed with this plugin (in folder DxCore). Is there way to insltall ONLY this avr-gcc without installing arduino (install it instead avr-gcc in system repositories)?

**Comment by SpenceKonde** [comment]
Apologies in advance for the arguably gratuitous profanity. I say arguably because I've yet to describe the process (at the end of the post) that I use to bodge together these toolchain packages to anyone other than my cat without hearing some from them - and even in the case of that sweet feline, I didn't like the tone of her meows ;-)

Here in Arduino circles, we always tell people that arduino installed from a system repository is no good (on many forums, the de-facto rule is "if you installed arduino via any repository, you will get no support untl you uninstall it and install the clean version to rule out problems there :-P )

All the post-2016 revolution boards for Arduino have always used a 7.3.0 based toolchain. If you're not willing to use Arduino or one of the tools like platform IO that uses Arduino cores, I wouldn't really know how to install it.

You can just download the tarball with the toolchain in it, but I don't think the directory tree is the same - it's arranged the way Arduino wants it arranged. What I'd probably do is install te 5.4.0 toolchain, and extract the azduino4 toolchain tarball and then compare the directory trees, (being really careful about nested folders of the same name - like, there is an avr/include and an avr/avr/include which have totally different files in them.....), and then copy all the 7.3.0 files ontop of their 5.4.0 counterparts. Because you need to replace:
Compiler binaries (you need 7.3.0 - arduino didn' move to that version for fun - they switch compiler versions only under extreme duress) including the .so files it loads. you need the device specs, and you need the includes. There are in excess of 1000 files at this point.... 
I don't know off hand which architecture raspi is, and it may depend on the version, but it's one of these two:
```
            {
              "size": "37878659",
              "checksum": "SHA-256:23becaaf746e8cf3b23ebcd261bf52d5eebcc094c3500fe8d533f0911b870c60",
              "host": "aarch64-linux-gnu",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-aarch64-pc-linux-gnu.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-aarch64-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "34220996",
              "checksum": "SHA-256:dacef5c90b38e6728e2cd161a54133b008352ff6f6513e67ab71f3117331c758",
              "host": "arm-linux-gnueabihf",
              "archiveFileName": "avr-gcc-7.3.0-atmel3.6.1-azduino4a-arm-linux-gnueabihf.tar.bz2",
              "url": "https://spencekondetoolchains.s3.amazonaws.com/avr-gcc-7.3.0-atmel3.6.1-azduino4a-arm-linux-gnueabihf.tar.bz2"
            },
```

There is no separate installer nor any plans to create one, for the simple reason that I absolutely fucking hate farting around with toolchains. Why do you think I use Arduino? :-P I mean FFS I develop on WINDOWS! I don't know how to use linux worth half a crap....  When I use linux. I don't use linux, stack overflow uses linux through my hands. 

Hell, do you think I actually know how to build this toolchain? I only know how to build for x64 linux on debian (and the result fails to actually link anything for mysterious reasons. but I don't care - because there are already 7.3.0 executables for all the Arduino platforms. It's just the support files I need (the precompiled .o files. plus device specs, headers and that kinda shit).... I fire up a machine on AWS (the build process takes almost half an hour on a beefy machine, and yes I am using all available threads, and it is using them, it just needs to compile an absurd amount of stuff. All the standard libraries get compiled for every class of AVR ever made). Then I take the resulting not-quite-functional x64 package back to a local linux system (and shut down the build machine, which costs far too much to leave running) where I also have a copy of the most recent official Arduino toolchain package (now 18 months old, and a very eventful 18 months for AVR releases). All the device specs, headers. .o's and so on get copied on-top of the ones in the old packages, while the compiler executables are left in place - I have a big text file that I copy and paste blocks of commands from..... then they get recompressed with a new name, and that's what I distribute. See, every day I spend farting around with toolchains and not enjoying myself is a day I could spend farting around with AVR assembly and C code, and loving it. When Microchiop starts shipping the DD-series parts, I'm going to have to do another build with the latest atpacks and repeat that unpleasant process :-P but that's far less work than figuring out how to crosscompile everything "the right way"
